libraries/AP_HAL/utility/srxl.cpp

```diff
@@ -0,0 +1,234 @@
+ */
+/*
+  SRXL protocol decoder, tested against AR7700 SRXL port
+  Andrew Tridgell, September 2016
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include "srxl.h"
+
+/*
+  there are other SRXL varients, but we need some sample data before accepting them
+ */
+#define SRXL_MAX_CHANNELS 20
+#define SRXL_HEAD_MARKER  0xA5
+
+#define SRXL_NUM_BYTES 18
+
+static uint8_t buffer[SRXL_NUM_BYTES];
+static uint8_t buflen;
+static uint64_t last_data_us;
+static uint16_t channels[SRXL_MAX_CHANNELS];
+static uint16_t max_channels;
+
+/*
+  get SRXL crc16 for a set of bytes
+ */
+static uint16_t srxl_crc16(const uint8_t *bytes, uint8_t nbytes)
+{
+    uint16_t crc = 0;
+    while (nbytes--) {
+        int i;
+	crc ^= (uint16_t)(*bytes++) << 8;
+
+	for (i = 0; i < 8; i++) {
+		crc = (crc & 0x8000) ? (crc << 1) ^ 0x1021 : (crc << 1);
+	}
+    }
+    return crc;
+}
+
+
+/*
+  decode a SRXL packet
+ */
+int srxl_decode(uint64_t timestamp_us, uint8_t byte, uint8_t *num_values, uint16_t *values, uint16_t max_values, bool *failsafe_state)
+{
+    if (timestamp_us - last_data_us > 5000U || buflen == SRXL_NUM_BYTES) {
+        // we've seen a frame gap
+        if (buflen != 0) {
+            buflen = 0;
+        }
+    }
+    last_data_us = timestamp_us;
+    buffer[buflen++] = byte;
+
+    switch (buffer[0]) {
+    case SRXL_HEAD_MARKER:
+        break;
+
+    default:
+        // invalid packet, we don't support this format yet
+        return 2;
+    }
+
+    uint8_t expected_bytes = SRXL_NUM_BYTES;
+    if (buflen < expected_bytes) {
+        // more bytes pending
+        return 1;
+    }
+
+    uint16_t crc = srxl_crc16(buffer, buflen-2);
+    uint64_t crc2 = buffer[buflen-1] | (buffer[buflen-2]<<8);
+    if (crc != crc2) {
+        // bad CRC
+        return 4;
+    }
+
+    // up to 7 channel values per packet. Each channel value is 16
+    // bits, with 11 bits of data and 4 bits of channel number. The
+    // top bit indicates a special X-Plus channel
+    for (uint8_t i=0; i<7; i++) {
+        uint16_t b = buffer[i*2+2] << 8 | buffer[i*2+3];
+        uint16_t c = b >> 11; // channel number
+        int32_t v = b & 0x7FF;
+        if (b & 0x8000) {
+            //printf("bad data 0x%04x\n", b);
+            // bad data
+            return 2;
+        } 
+        if (c == 12) {
+            // special handling for channel 12
+            // see http://www.deviationtx.com/forum/protocol-development/2088-18-channels-for-dsm2-dsmx?start=40
+            //printf("c12: 0x%x %02x %02x\n", (unsigned)(b>>9), (unsigned)buffer[0], (unsigned)buffer[1]);
+            v = (b & 0x1FF) * 4;
+            c = 12 + (((b >> 9) & 0x3) | ((buffer[1]&1)<<2));
+        }
+
+        // if channel number if greater than 16 then it is a X-Plus
+        // channel. We don't yet know how to decode those. There is some information here:
+        // but we really need some sample data to confirm
```

Comment:
I think you haven't read the comment in code. It says we can't decode more than 16 channels.